versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBStore.java

```diff
@@ -0,0 +1,392 @@
+        final LoadOp<?> loadOp = loadOps.get(i);
+        if (null == reads.get(i)) {
+          throw new NotFoundException(String.format("Unable to find requested ref with ID: %s", loadOp.getId()));
+        }
+
+        @SuppressWarnings("rawtypes") final ValueType type = loadOp.getValueType();
+        RocksSerDe.deserializeToConsumer(type, reads.get(i), loadOp.getReceiver());
+        loadOp.done();
+      }
+    }
+  }
+
+  @Override
+  public <C extends BaseValue<C>> boolean putIfAbsent(SaveOp<C> saveOp) {
+    final ColumnFamilyHandle columnFamilyHandle = getColumnFamilyHandle(saveOp.getType());
+    try {
+      final Transaction transaction = transactionDB.beginTransaction(new WriteOptions(), new OptimisticTransactionOptions());
+      // Get exclusive access to key if it exists.
+      final byte[] buffer = transaction.getForUpdate(new ReadOptions(), columnFamilyHandle, saveOp.getId().toBytes(), true);
+      if (null == buffer) {
+        transaction.put(columnFamilyHandle, saveOp.getId().toBytes(), RocksSerDe.serializeWithConsumer(saveOp));
+        transaction.commit();
+        return true;
+      }
+      // Id already exists.
+      return false;
+    } catch (RocksDBException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public <C extends BaseValue<C>> void put(SaveOp<C> saveOp, Optional<ConditionExpression> condition) {
+    final ColumnFamilyHandle columnFamilyHandle = getColumnFamilyHandle(saveOp.getType());
+
+    try (final Transaction transaction = transactionDB.beginTransaction(new WriteOptions(), new OptimisticTransactionOptions())) {
+      if (condition.isPresent()) {
+        final RocksBaseValue consumer = RocksSerDe.getConsumer(saveOp.getType());
+        final byte[] buffer = transaction.getForUpdate(new ReadOptions(), columnFamilyHandle, saveOp.getId().toBytes(), true);
+        if (null == buffer) {
+          throw new NotFoundException("Unable to load item with ID: " + saveOp.getId());
+        }
+        RocksSerDe.deserializeToConsumer(saveOp.getType(), buffer, consumer);
+
+        if (!consumer.evaluate(condition.get().accept(ROCKS_DB_CONDITION_EXPRESSION_VISITOR))) {
+          throw new ConditionFailedException("Condition failed during put operation");
+        }
+      }
+
+      transaction.put(columnFamilyHandle, saveOp.getId().toBytes(), RocksSerDe.serializeWithConsumer(saveOp));
+      transaction.commit();
+    } catch (RocksDBException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public <C extends BaseValue<C>> boolean delete(ValueType<C> type, Id id, Optional<ConditionExpression> condition) {
+    final ColumnFamilyHandle columnFamilyHandle = getColumnFamilyHandle(type);
+
+    try (final Transaction transaction = transactionDB.beginTransaction(new WriteOptions(), new OptimisticTransactionOptions())) {
+      final byte[] value = transaction.getForUpdate(new ReadOptions(), columnFamilyHandle, id.toBytes(), true);
+      if (value != null) {
```

Comment:
Can this be changed to match put() - so that it's value == null and the throw is early?

Reply:
Yes, done